### PR TITLE
treewide: drop DEV_DATA/DEV_CFG usage

### LINKED ISSUE #37616: Make access to device data and config consistent across drivers
**Is your enhancement proposal related to a problem? Please describe.**

Drivers usually need to access `data` and `config` fields from `struct device`. If one looks at `drivers/`, there seems to be a few different ways of accessing those fields. I think they can be classified in 3 _groups_:

1. Some drivers access `dev->config` and `dev->data` directly (I think this is the majority of drivers). In this case `void *` is not explicitly cast to its type.

https://github.com/zephyrproject-rtos/zephyr/blob/0f949caee12441dd4996017210d2d05971b3cf23/drivers/pwm/pwm_stm32.c#L266-L267

2. Some drivers provide helper macros like `DEV_CFG` or `DEV_DATA`. These macros usually perform an explicit cast to the actual config/data type.

https://github.com/zephyrproject-rtos/zephyr/blob/0f949caee12441dd4996017210d2d05971b3cf23/drivers/serial/uart_stm32.c#L45-L60

3. Some drivers provide helper inline functions like `get_dev_config` or `get_dev_data`. The return type of these functions is the actual config/data type, but in most (if not all) cases no cast is performed inside.

https://github.com/zephyrproject-rtos/zephyr/blob/0f949caee12441dd4996017210d2d05971b3cf23/drivers/serial/uart_nrfx_uarte.c#L175-L183

**Describe the solution you'd like**

I'd like to have consistency across drivers, that is, agree on a pattern, document it and convert all drivers. I think case (1) is the simplest and most common one, so I'd vote for that option.

**Describe alternatives you've considered**

N/A

**Additional context**

Any of the current solution (including explicit cast) violates MISRA 11.5 rule "A conversion should not be performed from pointer to void into pointer to object" (advisory category, though).

### PR DESCRIPTION
**Pull request by gmarull**
Make device driver data and configuration access more uniform accross drivers.

NOTE: a few other issues have been fixed as a result of the changes, e.g. many drivers dropping config `const` qualifier.

Fixes #37616

### CODE REVIEW

**Review by henrikbrixandersen — APPROVED** [review]
Thank you! This was on my TODO list 👍

**Review by nandojve — APPROVED** [review]
Nice!

File: drivers/clock_control/clock_control_esp32.c [hunk 1]
```diff
@@ -119,7 +117,7 @@ static int clock_control_esp32_get_rate(const struct device *dev,
 
 static int clock_control_esp32_init(const struct device *dev)
 {
-	struct esp32_clock_config *cfg = DEV_CFG(dev);
+	struct esp32_clock_config *cfg = dev->config;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/clock_control/clock_control_esp32.c]: fixed, thanks. Looks like many drivers were dropping `const` qualifier

File: drivers/dma/dma_mcux_lpc.c [hunk 2]
```diff
@@ -92,10 +90,11 @@ static void dma_mcux_lpc_irq_handler(const struct device *dev)
 static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 				  struct dma_config *config)
 {
+	const struct dma_mcux_lpc_config *config = dev->config;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/dma/dma_mcux_lpc.c]: good catch, fixed

**Review by jhedberg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Thank you, it was about time we cleaned this up.

**Pull request merged**